File: DLL.java

class Node
{
    public int data;
    public Node prev;
    public Node next;
}
class DoublyLL
{
    public Node first;
    public int size;
    DoublyLL()
    {
        this.first=null;
        this.size=0;
    }

    public void InsertFirst(int no)
    {
        Node newn=null;
        newn=new Node();
        newn.data=no;
        newn.prev=null;
        newn.next=null;
        if(first==null)
        {
            first=newn;
        }
        else
        {
             newn.next=first;
             first.prev=newn;
             first=newn;

        }
        this.size++;
    }

    public void InsertLast(int no)
    {
        Node newn=new Node();
        newn.data=no;
        newn.next=null;
        newn.prev=null;

        if(first==null)
        {
            first=newn;
        }
        else
        {
            Node temp=first;
            while(temp.next!=null)
            {
                temp=temp.next;
            }
            newn.prev=temp;
            temp.next=newn;
        }
        this.size++;
    }

    public void DeleteFirst()
    {
        if(size==0)
        {
            return;
        }
        else if(size==1)
        {
            first=null;
        }
        else
        {
            first=first.next;
        }
        this.size--;
    }
    public void DeleteLast()
    {
        if(size==0)
        {
            return;
        }
        else if(size==1)
        {
            first=null;
        }
        else
        {
            Node temp=first;
            while(temp.next.next!=null)
            {
                temp=temp.next;
            }
            temp.next=null;
        }
        this.size--;
    }
    public void Display()
    {
        Node temp=first;
        for(int i=1;i<=size;i++)
        {
            System.out.print("|"+temp.data+"|->");
            temp=temp.next;
        }
        System.out.print("null");
    }

    public void InsertAtPos(int no,int pos)
    {
        if((pos<0)||(pos>size+1))
        {
            return;
        }
        if(pos==1)
        {
            InsertFirst(no);
        }
        else if(pos==size+1)
        {
            InsertLast(no);
        }
        else
        {
            Node temp=first;
            Node newn=new Node();
            newn.data=no;
            newn.next=null;
            newn.prev=null;
            for(int i=1;i<pos-1;i++)
            {
                temp=temp.next;
            }
            newn.next=temp.next;
            newn.next.prev=newn;
            newn.prev=temp;
            temp.next=newn;
        }
        this.size++;
    }

    public void DeleteAtPos(int pos)
    {
        if((pos<0)||(pos>size))
        {
            return;
        }
        if(pos==1)
        {
            DeleteFirst();
        }
        else if(pos==size)
        {
            DeleteLast();
        }
        else
        {
            Node temp=first;
            for(int i=1;i<pos-1;i++)
            {
                temp=temp.next;
            }
            temp.next=temp.next.next;
            temp.next.prev=temp;
        }
        this.size--;
    }
    public int Count()
    {
        return size;
    }
}
class DLL
{
    public static void main(String arg[])
    {
        DoublyLL obj=new DoublyLL();
        obj.InsertFirst(51);
        obj.InsertFirst(21);
        obj.InsertFirst(11);
        obj.InsertLast(101);
        obj.DeleteFirst();
        obj.DeleteLast();
        obj.InsertAtPos(4, 2);
        obj.DeleteAtPos(2);
        obj.Display();
        int iRet=obj.Count();
        System.out.println("Count are:"+obj.Count());
    
    }
}

File: DLL.c
#include<stdio.h>
#include<stdlib.h>
typedef struct node
{
    int data;
    struct node*next;
    struct node*prev;
}NODE,*PNODE,**PPNODE;

void Display(PNODE Head)
{
    while(Head!=NULL)
    {
        printf("|%d|->",Head->data);
        Head=Head->next;
    }
    printf("NULL\n");
}

int Count(PNODE Head)
{
    int icnt=0;
    while(Head!=NULL)
    {
        icnt++;
        Head=Head->next;
    }
}

void InsertFirst(PPNODE Head,int no)
{
    PNODE newn=NULL;
    newn=(PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next=NULL;
    newn->prev=NULL;

    if(*Head==NULL)
    {
        *Head=newn;
    }
    else
    {
        newn->next=*Head;
        (*Head)->prev=NULL;
        *Head=newn;
    }
}

void InsertLast(PPNODE Head,int no)
{
    PNODE temp=*Head;
    PNODE newn=NULL;
    newn=(PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next=NULL;
    newn->prev=NULL;
    if(*Head==NULL)
    {
        *Head=newn;
    }
    else
    {
        while(temp->next!=NULL)
        {
            temp=temp->next;
        }
        temp->next=newn;
        newn->prev=temp;
    }
}
void DeleteFirst(PPNODE Head)
{
    if(*Head==NULL)
    {
        return;
    }
    else if((*Head)->next==NULL)
    {
        free(*Head);
        *Head=NULL;
    }
    else
    {
        *Head=(*Head)->next;
        free((*Head)->prev);
        (*Head)->prev=NULL;
    }
}

void DeleteLast(PPNODE Head)
{
    PNODE temp=*Head;
    if(*Head==NULL)
    {
        return;
    }
    else if((*Head)->next==NULL)
    {
        free(*Head);
        *Head=NULL;
    }
    else
    {
        while(temp->next!=NULL)
        {
            temp=temp->next;
        }
        temp->prev->next=NULL;
        free(temp);
    }
}

void InsertAtPos(PPNODE Head ,int no,int Pos)
{
    int size=0;
    PNODE temp=*Head;
    PNODE newn=NULL;
    int i=0;
    size=Count(*Head);
    if((Pos<1)||(Pos>size+1))
    {
        printf("Enter valid position\n");
    }
    if(Pos==1)
    {
        InsertFirst(Head,no);
    }
    else if(Pos==size+1)
    {
        InsertLast(Head,no);
    }

    else
    {
        newn=(PNODE)malloc(sizeof(NODE));
        newn->data=no;
        newn->next=NULL;
        newn->prev=NULL;
        for(i=1;i<Pos-1;i++)
        {
            temp=temp->next;
        }
        newn->next=temp->next;
        temp->next->prev=newn;
        temp->next=newn;
        newn->prev=temp;
    }
}

void DeleteAtPos(PPNODE Haed,int Pos)
{
    PNODE temp=*Haed;
    int size=Count(*Haed);
    int i=0;
    if((Pos<1)||(Pos>size))
    {
        printf("Enter valid");
        return;
    }
    if(Pos==1)
    {
        DeleteFirst(Haed);
    }
    else if(Pos==size)
    {
        DeleteLast(Haed);
    }
    else
    {
        for(i=1;i<Pos-1;i++)
        {
            temp=temp->next;
        }
        temp->next=temp->next->next;
        free(temp->next->prev);
        temp->next->prev=temp;
    }


}

int main()
{
    PNODE first=NULL;
    int Choice=1;
    int value=1,Pos=0,iRet=0;
    while(Choice!=0)
    {
        printf("\n------------------------------------------------\n");
        printf("Enter desired operation you want perform in DLL:\n");
        printf("1:Insert First:\n");
        printf("2:Insert Last\n");
        printf("3:Insert At Position\n");
        printf("4:Delete First\n");
        printf("5:Delete Last\n");
        printf("6:Delete At Position\n");
        printf("7:Display DLL\n");
        printf("8:Count element DLL\n");
        printf("0:Thank You!!!!!!!!!\n");
        printf("--------------------------------------------------\n");
        scanf("%d",&Choice);
        printf("----------------------------------------------------\n");

        switch(Choice)
        {
            case 1:
            printf("Enter Data\n");
            scanf("%d",&value);
            InsertFirst(&first,value);
            break;

            case 2:
            printf("Enter Data\n");
            scanf("%d",&value);
            InsertLast(&first,value);
            break;
            
            case 3:
            printf("Enter Data\n");
            scanf("%d",&value);
            printf("Enter Position\n");
            scanf("%d",&Pos);
            InsertAtPos(&first,value,Pos);
            break;

            case 4:
            DeleteFirst(&first);
            break;

            case 5:
            DeleteLast(&first);
            break;

            case 6:
            printf("Enter Position\n");
            scanf("%d",&Pos);
            DeleteAtPos(&first,Pos);
            break;

            case 7:
            Display(first);
            break;

            case 8:
            iRet=Count(first);
            printf("Count Are:%d\n",iRet);
            break;

            case 0:
            printf("!!! Thank You !!!\n");
            break;

            default:
            printf("Enter valid choice\n");
            break;
        }
    }
    return 0;
}